pkg/services/ngalert/schedule/schedule.go

```diff
@@ -315,11 +317,21 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 	evalDuration := sch.metrics.EvalDuration.WithLabelValues(orgID)
 	evalTotalFailures := sch.metrics.EvalFailures.WithLabelValues(orgID)
 
-	clearState := func() {
-		states := sch.stateManager.ResetStateByRuleUID(grafanaCtx, key)
-		expiredAlerts := FromAlertsStateToStoppedAlert(states, sch.appURL, sch.clock)
-		if len(expiredAlerts.PostableAlerts) > 0 {
-			sch.alertsSender.Send(key, expiredAlerts)
+	clearState := func(ctx context.Context, err error) {
+		// clean up the state only if the reason for stopping the evaluation loop is that the rule was deleted/paused
+		isRuleDeleted := errors.Is(err, ErrRuleDeleted)
+		isRulePaused := errors.Is(err, ErrRulePaused)
+		if isRuleDeleted || isRulePaused {
+			var states []*state.State
+			if isRuleDeleted {
+				states = sch.stateManager.ResetStateByDeletedRuleUID(ctx, rule)
```

Comment:
I would recommend getting the the rule from the registry here using the rule key